tests/ComTests.cpp

```diff
@@ -2819,4 +2822,127 @@ TEST_CASE("StreamTests::Saver", "[com][IStream]")
+        static_assert(std::is_same_v<deduced_next_t::Result, LPITEMIDLIST>);
+
+        using traits_t = wil::details::com_enumerator_traits<IEnumIDList>;
+        static_assert(std::is_same_v<traits_t::Result, LPITEMIDLIST>);
+        static_assert(std::is_same_v<traits_t::smart_result, LPITEMIDLIST>);
+    }
+    {
+        using iterator_t = wil::iterator<IEnumIDList>;
+        iterator_t it(nullptr);
+        static_assert(std::is_same_v<LPITEMIDLIST&, decltype(*it)>);
+    }
+    {
+        // mock 0-size iterator
+        struct IEnumMuffins : IUnknown
+        {
+            // IUnknown
+            HRESULT __stdcall QueryInterface(REFIID, void**) noexcept { return E_NOINTERFACE; }
+            ULONG __stdcall AddRef() noexcept { return 0; }
+            ULONG __stdcall Release() noexcept { return 0; }
+            // IEnumMuffins
+            HRESULT __stdcall Next(ULONG, int32_t*, ULONG*) noexcept { return S_FALSE; }
+        };
+        static_assert(std::is_same_v<wil::details::com_enumerator_next_traits<decltype(&IEnumMuffins::Next)>::Result, int32_t>);
+        static_assert(std::is_same_v<wil::details::com_enumerator_next_traits<decltype(&IEnumMuffins::Next)>::Interface, IEnumMuffins>);
+        static_assert(std::is_same_v<wil::details::com_enumerator_traits<IEnumMuffins>::Result, int32_t>);
+        static_assert(std::is_same_v<wil::details::com_enumerator_traits<IEnumMuffins>::smart_result, int32_t>);
+
+        auto count = 0;
+        auto muffins = IEnumMuffins{};
+        for (auto muffin : wil::make_range(&muffins))
+        {
+            REQUIRE(muffin == 0);
+            count++;
+            break;
+        }
+        REQUIRE(count == 0);
+
+        struct IEnumMuffinsCOM : IUnknown
+        {
+            // IUnknown
+            HRESULT __stdcall QueryInterface(REFIID, void**) noexcept { return E_NOINTERFACE; }
+            ULONG __stdcall AddRef() noexcept { return 0; }
+            ULONG __stdcall Release() noexcept { return 0; }
+            // IEnumMuffinsCOM
+            HRESULT __stdcall Next(ULONG, IUnknown**, ULONG*) noexcept { return S_FALSE; }
+        };
+        static_assert(std::is_same_v<wil::details::com_enumerator_next_traits<decltype(&IEnumMuffinsCOM::Next)>::Result, IUnknown*>);
+        static_assert(std::is_same_v<wil::details::com_enumerator_next_traits<decltype(&IEnumMuffinsCOM::Next)>::Interface, IEnumMuffinsCOM>);
+        static_assert(std::is_same_v<wil::details::com_enumerator_traits<IEnumMuffinsCOM>::Result, IUnknown*>);
+        static_assert(std::is_same_v<wil::details::com_enumerator_traits<IEnumMuffinsCOM>::smart_result, wil::com_ptr<IUnknown>>);
+        auto muffinsCOM = IEnumMuffinsCOM{};
+        count = 0;
+        for (auto muffin : wil::make_range(&muffinsCOM))
+        {
+            REQUIRE(muffin == nullptr);
+            count++;
+            break;
+        }
+        REQUIRE(count == 0);
+        
+    }
+    {
+#if (NTDDI_VERSION >= NTDDI_VISTA)

```

Comment:
This is already in an ifdef cpp17 block.  Can you even compile pre-Vista headers in C++ 17 mode?

Reply:
not sure but i don't think we should tie together/assume lang version x SDK version. One check is for one thing, the other is for another, if the combination of C++17 w/pre-Vista is invalid, there's no harm (this check just makes it clear that this API was added in Vista)